modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -37,12 +48,31 @@ object GlobalL0Service {
     l0GlobalSnapshotClient: L0GlobalSnapshotClient[F],
     globalL0ClusterStorage: L0ClusterStorage[F],
     lastGlobalSnapshotStorage: LastSnapshotStorage[F, GlobalIncrementalSnapshot, GlobalSnapshotInfo],
-    singlePullLimit: Option[PosLong]
+    singlePullLimit: Option[PosLong],
+    maybeRealignPeerIds: Option[NonEmptySet[PeerId]] = none
   ): GlobalL0Service[F] =
     new GlobalL0Service[F] {
-
       private val logger = Slf4jLogger.getLogger[F]
 
+      def realignToMajority(lastSnapshot: Hashed[GlobalIncrementalSnapshot]): F[Unit] =
+        maybeRealignPeerIds.traverse_ { realignPeers =>
+          for {
+            peers <- realignPeers.toNonEmptyList
+              .traverse(globalL0ClusterStorage.getPeer)
+              .map(_.toList.flatten)
+
+            majorityHash <- peers
+              .parTraverseN(10)(p => l0GlobalSnapshotClient.getHash(lastSnapshot.ordinal).run(p))
+              .map(_.flatten)
+              .map(pickMajority[List, Hash])
+
+            isAligned = majorityHash.fold(false)(lastSnapshot.hash === _)
```

Comment:
For the case where majority hash wasn't discovered, do we want to treat it as misalignment?

Reply:
Continue on meaning trying again periodically for consecutive pulls until we successfully determine the majority for given ordinal based on priority peers or defaulting to random pick from peers in the storage? If you mean the second one then we would be breaking the assumption of only pulling pre-validated majority snapshots and hence we could fork. Also in a situation of a network restart none of priority nodes may be present so they won't respond and if we treat no response as no basis for halting but actually we will continue to pull snapshots from other nodes we will fork and this whole functionality will make no sense in case of a rollback.